### ISSUE TITLE
Class names randomly changed from `ie` to `y` suffix

### ISSUE THREAD

**Issue by mindplay-dk** [quicktype/quicktype#1867]
In my generated classes, some type-names are randomly changed from ending in `ie` to ending in `y`.

Uh, yeah, I know - but look:

![image](https://user-images.githubusercontent.com/103348/149935320-3f711902-9e26-4f3f-a3d5-0123e714c89d.png)

I swear, this is not a prank - some types (not all of them) in my input JSON schema are somehow getting renamed.

It does not appear to have anything to do with the output format - this happens with at least both TypeScript and C# output.

Again, this is not a prank - but maybe some joker introduced this into your codebase as a prank?

I gotta say, it is kind of funny. 😉

What could possibly cause this?

**Comment by mindplay-dk** [comment]
Uh, and here it is now, apparently cutting a class-name in half:

![image](https://user-images.githubusercontent.com/103348/149966180-03591291-6fdc-4850-a74a-5ce1c3fd4bd8.png)

Come on now, guys - joke's over!

**Comment by eliassorensen** [comment]
Looks like it could be quicktype's use of `pluralize`. Doesn't seem like this is configurable in quicktype? 

Your second comment, I cannot at all see how/why this should happen. In general, I think the way pluralize is used here is flawed.

**Comment by mindplay-dk** [comment]
Yup, looks like `JSONSchemaInput` unconditionally attempts to singularize type-names - here:

https://github.com/quicktype/quicktype/blob/daffe53101c4d329c0647fb8310aded585e3808e/src/quicktype-core/input/JSONSchemaInput.ts#L662

And here:

https://github.com/quicktype/quicktype/blob/daffe53101c4d329c0647fb8310aded585e3808e/src/quicktype-core/input/JSONSchemaInput.ts#L790

Lucky for me, I'm using the API rather than the CLI - so for my project, I should be able to reconfigure `pluralize`.

For people using the CLI, not so much. There probably should be an option to bypass name mangling.

Though, to begin with, I wonder, why is it, by default, we don't "trust" the names in JSON schema? Whether those names "look pretty" as class-names or not, there are those of us who would prefer consistency with the names in the original API specification - I personally don't think this behavior makes sense as a default for JSON schema.

(It might not make sense as a default under any circumstances, given how poorly `pluralize` works in this practical case.)

**Comment by mindplay-dk** [comment]
Ah, I understand now - for nested types in JSON schema that don't *have* type-names, it *has* to come up with a name, in order to generate additional classes in e.g. C# and other languages that (unlike e.g. TypeScript) do not support structural subtyping.

So this is a feature, not a bug.

The issue remains though, that this doesn't actually work well.

Might we consider adding a warning or something? When it has to make up missing names on the fly.

If there was a warning, and an option to manually inject singular/plural word pairs, that would seem a lot more robust.

Having to break into third-party libraries and rely on global state and side-effects ain't pretty. We should not be coupled to `pluralize` on the consumer side - it should be an implementation detail in the library, I think.

**Comment by schani** [comment]
We do not take feature requests anymore. Please consider making a pull request instead. We encourage the use of coding agents, but please use at least an Opus-class model.
